### PR TITLE
Add Osprey (DIA) search-engine integration: Koina library, entrapment/FDRBench validation, finetune, and installer

### PR DESCRIPTION
**Pull request by maccoss**
Consolidates the Osprey (DIA search-engine) integration into Carafe, validated
end-to-end on real data. `main` is a clean fast-forward superset of
`Noble-Lab:main` (34 commits, no conflicts).

**Supersedes #8, #9, and #10** — those scoped branches diverged as the work
continued (the finetune and decoy/FDRBench fixes below landed only on `main`),
so they're folded into this single reviewable PR.

## Highlights

**Osprey search-engine integration**
- New "Osprey" DIA search engine wired through the GUI (Workflows 4 & 5):
  search → finetune → build a new library → re-search the project files.
- `OspreyBinaryResolver` resolves the executable, preferring the
  installer-bundled build over a stale saved path.
- Tracks the upstream `OspreySharp` → `Osprey` rename (namespaces, `Osprey.exe`).

**Initial library generation**
- Koina (remote) and Carafe-local AlphaPepDeep predictors produce the starting
  DIA-NN library; `KoinaLibraryGenerator` enumerates peptidoforms/charges and
  writes a DIA-NN TSV.

**Entrapment + FDRBench (FDP validation)**
- Entrapment peptides go **only** in the library FASTA, never the training
  FASTA (`OspreyFastaPlanner`); a 5-column FDRBench pairing manifest +
  `--fdrbench` input bundle drive entrapment-based FDP measurement
  (`OspreyFdrBenchPlanner`).
- Decoys are generated the Osprey way (C-term-preserving reverse, cycle on
  collision), marked correctly, and made unique against both targets and
  entrapments.

**Finetune correctness fixes**
- Don't hard-exit when a peak apex falls between scans (unit-resolution data).
- Read the synthesized `ms2index` (true MS2 scan ordinal) instead of the Osprey
  blib's placeholder `MS2.Scan=0` — without this the MS2 model trained on zero
  valid PSMs and crashed.
- **Multi-file (folder) training**: fixed a crash in the Osprey finetune step
  when training on a folder of mzMLs, and emit the `Precursor.Id` column
  (`Modified.Sequence` + `Precursor.Charge`) that the Osprey report was missing —
  the multi-run "best" merge does `hIndex.get("Precursor.Id")` and NPE'd without
  it. `Precursor.Id` is now built identically to Carafe's own Skyline→DIA-NN
  converter so both report sources merge the same way.

**CI / installer**
- GitHub Actions workflow + script build a Windows MSI bundling Osprey
  (`Osprey.exe`); folder-publish (not single-file) so the SQLite `.blib` writer
  works; stage the `.blib` locally then move (NAS/SMB can't host the SQLite
  write); actions pinned off deprecated Node 20; vendored patched
  compomics-utilities jar for clean-machine builds.
- A second GitHub Actions workflow runs the unit-test suite on every push; the
  MSI version and build stamp are derived from the Maven project version (single
  source of truth) so a pom bump can't drift from a hard-coded value.

**Docs**: an end-to-end Osprey workflow document describing how the GUI calls
each tool.

## Tests
New unit tests: `OspreyFastaPlannerTest`, `OspreyFdrBenchPlannerTest`,
`EntrapmentFastaGearTest` (Osprey-style decoys), `PSMConfigTest` (ms2index
re-point regression), `KoinaLibraryGeneratorTest` (end-to-end via a stubbed
client), `OspreyBinaryResolverTest`, `AIGearMsFileResolutionTest` (multi-file
resolution), and `OspreyBlibReaderTest` (incl. the `Precursor.Id` regression).
Full suite green (105 tests, 0 failures).

## Validation
End-to-end on Stellar HeLa entrapment data: at q = 0.01, **combined FDP 0.98% /
paired FDP 0.88%** — on/below the unity line, i.e. the reported FDR is accurate
(slightly conservative).

## Dependency
The Osprey side (`--fdrbench` output + a library-cache staleness fix) has
**merged into ProteoWizard/pwiz `master`** — ProteoWizard/pwiz#4337 (`--fdrbench`),
ProteoWizard/pwiz#4338 (stale `.libcache`), plus the base_id decoy-pairing fix
ProteoWizard/pwiz#4347. The bundled MSI now builds Osprey from upstream `master`;
there is no longer a fork dependency for the Osprey side.

## Test coverage — driven by real integration findings

Many of the tests here exist because we hit concrete failures bringing the Osprey integration up
end-to-end. They're regression guards for bugs we actually encountered, not speculative coverage;
the suite grew **69 → 105 tests** over this work:

- **`ms2index` / "Spectrum not found"** — the finetune trained on *zero* valid MS2 PSMs until the
  Osprey-blib → training-input column handoff was fixed. Guarded by `PSMConfigTest`, and the
  precursor → MS2-scan matching was extracted out of `AIGear` into a unit-tested `ApexMatcher`.
- **Multi-file (folder) training NPE** — training on a folder of mzMLs crashed in the multi-run
  "best" merge because the Osprey report omitted `Precursor.Id`. `OspreyBlibReaderTest` now pins
  that the column is present and equals `Modified.Sequence` + `Precursor.Charge`, and
  `AIGearMsFileResolutionTest` covers the multi-file path.
- **The dia/ spectrum layer** (previously 0% covered — where the "spectrum not found" lookups live):
  `DIASpectrumTest` parses a tiny synthetic mzML through the real MSFTBX path and checks `DIAMeta`
  parsing + `DIAIndex.get_spectrum_by_scan`; `DIAMapTest` pins isolation-window membership; and
  `IsolationWindowTest` pins the bucket-key rounding the decoy pairing depends on.
- **Stale library cache** (SEA-AD: manifest decoy pairing collapsed to ~0.4%) — the window
  bucketing it hinged on is now covered by the dia/ tests above.
- **Locale / wrong-mass / unrecognized-mod / null-input** edge cases from the Koina and
  `OspreyBlibReader` work — each now has a focused test (including the unknown-mod *skip*).
- Foundational gaps the bugs exposed: `DBGearTest` (precursor m/z), the reuse-signature
  fingerprint, and Koina response-size validation.

These joined the fixes from the automated-review pass (Copilot + CodeRabbit). Full suite: 105 tests,
0 failures.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added full Osprey integration for workflows 4/5 in the GUI and CLI, including entrapment support, Koina-based library generation, and end-to-end orchestration.
  * Added tools for comparing spectral library predictions and building Osprey binaries/installers (including Windows MSI automation).
* **Bug Fixes**
  * Improved multi-file MS2 apex matching and made workflow reuse/skip more reliable by tracking step inputs.
  * Hardened process stopping/cleanup and improved behavior when mzML/apex resolution is incomplete.
* **Documentation**
  * Expanded README and added a dedicated Osprey end-to-end workflow guide (plus vendored dependency notes).
* **Tests**
  * Added extensive unit/integration coverage for Osprey, Koina, mzML reading, and new reconciliation/utilities.

### CODE REVIEW

File: scripts/build_osprey.sh [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    case "${os}" in
+        Linux)  echo "linux-x64" ;;
+        Darwin)
+            case "${arch}" in
+                arm64) echo "osx-arm64" ;;
+                *)     echo "osx-x64" ;;
+            esac ;;
+        MINGW*|MSYS*|CYGWIN*) echo "win-x64" ;;
+        *)      echo "linux-x64" ;;
+    esac
```
[the remark this answers is not in the document]
Comment (maccoss) [reply scripts/build_osprey.sh:49]: Fixed in a38b7aa — the MSBuild Platform is now derived from the RID architecture (x64/ARM64/x86) in both `build_osprey.sh` and `build_osprey.bat`, so `osx-arm64`/`win-arm64` publish with a matching Platform.

File: scripts/build_osprey.sh [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+#
+# Environment overrides:
+#   OSPREY_CSPROJ  Path to Osprey.csproj
+#                  (default: ~/GitHub-Repo/ProteoWizard/pwiz/pwiz_tools/Osprey/Osprey/Osprey.csproj)
+#   OUTPUT_ROOT    Where to stage per-RID output
+#                  (default: <repo>/target/osprey)
+#
+# Examples:
+#   scripts/build_osprey.sh                # build the default RID for this host
+#   scripts/build_osprey.sh win-x64 linux-x64 osx-arm64
```
[the remark this answers is not in the document]
Comment (maccoss) [reply scripts/build_osprey.sh]: Fixed in a38b7aa — `osx-x64` added to the usage examples.

File: scripts/compare_library_predictions.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    with path.open() as f:
+        reader = csv.DictReader(f, delimiter="\t")
+        for row in reader:
+            if row.get("StrippedPeptide") != peptide:
+                continue
+            if int(float(row.get("PrecursorCharge", 0))) != charge:
+                continue
+            key = (row["FragmentType"], int(row["FragmentNumber"]), int(row["FragmentCharge"]))
+            frags[key] = float(row["RelativeIntensity"])
+    return frags
```
[the remark this answers is not in the document]
Comment (maccoss) [reply scripts/compare_library_predictions.py]: Fixed in a38b7aa — `load_fragments` now also filters by `ModifiedPeptide` (new `--modified-peptide`) and fails (`SystemExit`) when a stripped-sequence+charge still matches multiple peptidoforms, so the cosine is never computed from a mixed spectrum.

File: src/main/java/gui/CarafeGUI.java [hunk 4]
```diff
@@ -349,6 +416,9 @@ private void initComponents() {
         tabbedPane.addTab("Training Data Generation", wrapInScrollPane(createTrainingDataPanel()));
         tabbedPane.addTab("Model Training", wrapInScrollPane(createModelTrainingPanel()));
         tabbedPane.addTab("Library Generation", wrapInScrollPane(createLibraryGenerationPanel()));
+        tabbedPane.addTab("Osprey", wrapInScrollPane(createOspreyPanel()));
+        // Console is added last (rightmost); code switches to it via indexOfTab("Console"),
+        // so its position is not hardcoded.
```
[the remark this answers is not in the document]
Comment (maccoss) [reply src/main/java/gui/CarafeGUI.java:463]: Fixed in a38b7aa — `setInputsFrozen` now freezes every tab except the Console (via `WorkflowTabs.freezableTabIndices`) rather than a fixed count of four, so the Osprey tab is locked during runs. Unit-tested in `WorkflowTabsTest`.

File: src/main/java/gui/CarafeGUI.java [hunk 5]
```diff
@@ -4766,6 +5116,234 @@ private void runCarafe() {
...
+                        JOptionPane.showMessageDialog(this, "No valid project MS files (mzML/.d) found.",
+                                "Input Required", JOptionPane.WARNING_MESSAGE);
+                        setInputsFrozen(false);
+                        return;
+                    }
+                }
+
+                boolean isTimsTOF = trainMs.stream().anyMatch(p -> p.toLowerCase().endsWith(".d"));
+                String trainMsInput = trainMs.size() == 1 ? trainMs.getFirst()
+                        : new File(trainMs.getFirst()).getParent();
```
[the remark this answers is not in the document]
Comment (maccoss) [reply src/main/java/gui/CarafeGUI.java:5349]: Fixed in a38b7aa — the Osprey training path now rejects files from different folders (`OspreyMsInputPlanner.allInSameDirectory`) with a clear message, matching the DIA-NN guard. Unit-tested in `OspreyMsInputPlannerTest`.

File: src/main/java/gui/CarafeGUI.java [hunk 6]
```diff
@@ -5052,6 +5665,262 @@ private int runSingleCommand(CmdTask task, String pythonPath) throws Exception {
...
+                    pool.shutdown();
+                    for (java.util.concurrent.Future<?> fut : futures) {
+                        try {
+                            fut.get();
+                        } catch (Exception e) {
+                            failed.set(true);
+                        }
+                    }
+                    if (failed.get()) {
+                        break;
```
[the remark this answers is not in the document]
Comment (maccoss) [reply src/main/java/gui/CarafeGUI.java:6085]: Fixed in a38b7aa — on the first lane failure, `abortRunningLanes()` terminates the outstanding tracked subprocesses (`activeProcesses`) via `ProcessUtils.terminateAll`, so a single failure no longer lets the other lanes run to completion.

File: .github/workflows/build-installer.yml [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+  contents: write
+
+jobs:
+  build-msi:
+    runs-on: windows-latest
+
+    env:
+      # Defaults used when the workflow is triggered by a tag (no manual inputs).
+      OSPREY_REPO: ${{ github.event.inputs.osprey_repo || 'ProteoWizard/pwiz' }}
+      OSPREY_REF: ${{ github.event.inputs.osprey_ref || 'master' }}
```
[the remark this answers is not in the document]
Comment (maccoss) [reply .github/workflows/build-installer.yml:43]: Fixed in a38b7aa — release (v* tag) builds now pin Osprey to the `RELEASE_OSPREY_REF` repository variable and fail fast if it is unset, instead of falling back to master tip. Manual dispatch is unchanged. NOTE: set `RELEASE_OSPREY_REF` to a reviewed pwiz commit/tag before the next tagged release.

File: .github/workflows/test.yml [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  pull_request:
+  workflow_dispatch:
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v5
```
[the remark this answers is not in the document]
Comment (maccoss) [reply .github/workflows/test.yml:18]: Fixed in a38b7aa — `persist-credentials: false` added to the checkout in `test.yml` (and the `build-installer.yml` checkout too).

File: src/main/java/db/PairingManifestReconciler.java [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+                    // Fail loudly rather than silently dropping a peptide/group from a corrupt manifest.
+                    throw new IOException("malformed manifest " + manifestIn + " at line " + lineNo
+                            + ": expected at least " + (iPair + 1) + " tab-separated columns, found "
+                            + c.length);
+                }
+                r.rowsIn++;
+                groups.computeIfAbsent(c[iPair], k -> new ArrayList<>())
+                        .add(new Row(c[iSeq], c[iDecoy], c[iProteins], c[iType]));
+            }
+        }
```
[the remark this answers is not in the document]
Comment (maccoss) [reply src/main/java/db/PairingManifestReconciler.java:145]: Fixed in a38b7aa — the malformed-row guard now checks against the max of all resolved column indices, not just `peptide_pair_index`, so a reordered header is handled. Covered by `handlesReorderedHeader` and `failsFastOnShortRowWithReorderedHeader`.

File: .github/workflows/build-installer.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (maccoss) [reply .github/workflows/build-installer.yml:43]: Update (051735e): per maintainer decision, release (v* tag) builds intentionally track Osprey **master** rather than requiring a pinned ref — the fail-fast guard was reverted. Manual dispatch can still override the ref. Resolving as a deliberate "track master" choice.

File: src/main/java/util/ReuseSignature.java [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+                if (f.isDirectory()) {
+                    File[] kids = f.listFiles(File::isFile);
+                    if (kids != null) {
+                        for (File k : kids) {
+                            lines.add(fingerprintFile(k));
+                        }
+                    }
+                } else {
+                    lines.add(fingerprintFile(f));
+                }
```
[the remark this answers is not in the document]
Comment (maccoss) [reply src/main/java/util/ReuseSignature.java:120]: Fixed in 040aee4. `fingerprintInputs()` now walks the whole tree with `Files.walk` + `Files::isRegularFile`, so directory-backed inputs like Bruker `.d` fingerprint every nested file; an `IOException` emits a `DIR_WALK_FAILED` marker so a failed walk invalidates rather than silently matching. See `ReuseSignature.java:109-120`.

File: src/main/java/util/ReuseSignature.java [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+        lines.sort(String::compareTo);
+        return String.join("\n", lines);
+    }
+
+    private static String fingerprintFile(File f) {
+        if (f.exists()) {
+            return "in:" + f.getName() + "|" + f.length() + "|" + f.lastModified();
+        }
+        return "in:" + f.getName() + "|MISSING";
+    }
```
[the remark this answers is not in the document]
Comment (maccoss) [reply src/main/java/util/ReuseSignature.java:136]: Fixed in 040aee4. `fingerprintFile()` now keys on `f.getAbsolutePath()` instead of the basename, so same-named files in different folders (and directory-expanded files) no longer collide. See `ReuseSignature.java:127-136`.

File: src/main/java/util/ProcessUtils.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+     * them to exit. Used by the Stop button and the JVM shutdown hook so that closing the GUI
+     * mid-run never leaves orphaned converter processes behind. Best effort: null/dead entries and
+     * any per-process errors are ignored.
+     *
+     * @param processes the processes to terminate (e.g. {@code cmd /c msconvert ...} wrappers whose
+     *                  child is the real converter)
+     */
+    public static void terminateAll(Collection<Process> processes) {
+        List<Process> list = new ArrayList<>(processes);
+        // First pass: kill descendants (the actual converter spawned by the shell), then the
```
[the remark this answers is not in the document]
Comment (maccoss) [reply src/main/java/util/ProcessUtils.java:31]: Fixed in 4eadb84. `terminateAll()` returns early on a null collection, matching the "best effort / null inputs ignored" contract in the Javadoc. See `ProcessUtils.java:26-29`.

File: src/main/java/koina/KoinaLibraryGenerator.java [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+        for (int oi = 0; oi < limit; oi++) {
+            int idx = order[oi];
+            float[] f = frags.get(idx);
+            String[] meta = fragMeta.get(idx);
+            float relInt = f[1] / maxInt;
+            w.write(p.form.diann + "\t" + p.form.stripped + "\t"
+                    + String.format("%.5f", p.mz) + "\t" + p.charge + "\t"
+                    + String.format("%.4f", irt) + "\t" + p.proteinId + "\t" + p.decoy + "\t"
+                    + String.format("%.5f", f[0]) + "\t" + String.format("%.6f", relInt) + "\t"
+                    + meta[0] + "\t" + (int) f[2] + "\t" + (int) f[3] + "\t" + meta[1] + "\n");
```
[the remark this answers is not in the document]
Comment (maccoss) [reply src/main/java/koina/KoinaLibraryGenerator.java:279]: Fixed in 4eadb84. Every numeric field in `writePrecursor()` now formats with `Locale.ROOT`, so a comma-decimal locale can no longer emit unparseable TSV. See `KoinaLibraryGenerator.java:276-278`.

File: src/main/java/koina/KoinaClient.java [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+        JSONObject root = JSON.parseObject(json);
+        float[] intensities = floatOutput(root, "intensities");
+        float[] mz = floatOutput(root, "mz");
+        String[] annotation = stringOutput(root, "annotation");
+        if (intensities == null || mz == null || annotation == null) {
+            throw new IllegalStateException("Koina MS2 response missing intensities/mz/annotation: "
+                    + truncate(json));
+        }
+        int f = intensities.length / n;
+        List<Ms2> out = new ArrayList<>(n);
```
[the remark this answers is not in the document]
Comment (maccoss) [reply src/main/java/koina/KoinaClient.java:167]: Fixed in 040aee4. `parseMs2Response()` now validates up front and fails fast with the offending JSON truncated into the message: null outputs, the three arrays differing in length, and `intensities.length % n != 0` (including `n <= 0`) each throw `IllegalStateException` before the per-precursor split. See `KoinaClient.java:147-162`.

File: src/main/java/ai/AIGear.java [hunk 15]
```diff
@@ -870,6 +892,127 @@ public static void main(String[] args) throws ParseException, IOException {
...
+                System.err.println("-build_entrapment_fasta requires an input protein FASTA via -db");
+                System.exit(1);
+            }
+            EntrapmentFastaGear.Config efc = new EntrapmentFastaGear.Config();
+            efc.inputFasta = input_fasta;
+            efc.outputFasta = cmd.getOptionValue("build_entrapment_fasta");
+            efc.manifest = cmd.getOptionValue("manifest"); // null if not provided
+            efc.addEntrapment = cmd.hasOption("entrapment");
+            efc.addDecoys = !cmd.hasOption("no_decoys");
+            efc.applyMzFilter = cmd.hasOption("mz_filter");
```
[the remark this answers is not in the document]
Comment (maccoss) [reply src/main/java/ai/AIGear.java:935]: Fixed in 4eadb84. The `-build_entrapment_fasta` path now propagates the flag into `EntrapmentFastaGear.Config` (`efc.decoyPrefix = cmd.getOptionValue("decoy_prefix")`), so the generated FASTA labels decoys consistently with the rest of the pipeline. See `AIGear.java:926-929`.

File: README.md [hunk 16]
```diff
@@ -153,6 +153,43 @@ For comparison purposes, Workflow 3 also provides the option shown below to perf
...
+```bash
+# Linux/macOS
+scripts/build_ospreysharp.sh            # current platform
+scripts/build_ospreysharp.sh win-x64 linux-x64 osx-arm64
+
+# Windows
+scripts\build_ospreysharp.bat win-x64
+```
+
+This stages `OspreySharp(.exe)` under `target/osprey/<rid>/`. Place the appropriate `<rid>/` directory beside the Carafe2 jar as `osprey/<rid>/` (or under `~/.carafe/osprey/<rid>/`); Carafe2 locates it automatically at runtime. You can also point Carafe2 at a custom OspreySharp build via its path preference. Override the OspreySharp source location with the `OSPREY_CSPROJ` environment variable if your `pwiz` checkout is elsewhere.
```
[the remark this answers is not in the document]
Comment (maccoss) [reply README.md]: Fixed in 5747751 (and 7a8fc47). The scripts are now `scripts/build_osprey.{sh,bat}` and README/installer text refers to `Osprey` / `Osprey.exe`, matching what is actually published and what `resolveOspreyBinary()` looks for. `grep -ri ospreysharp` over the tree returns nothing.

File: docs/ospreysharp-end-to-end-workflow.md [hunk 17]
```diff
@@ -0,0 +1,483 @@
+# The OspreySharp end-to-end workflow
+
+This document describes, step by step, what Carafe2 does when it runs a spectral-library
+workflow with **OspreySharp** as the search engine — from the input protein FASTA all the way
+to the final fine-tuned spectral library (and, in the end-to-end variant, the search of your
+project files with that library).
```
[the remark this answers is not in the document]
Comment (maccoss) [reply docs/ospreysharp-end-to-end-workflow.md]: Fixed in 5747751 (and 7a8fc47). The doc is now `docs/osprey-end-to-end-workflow.md` and uses `Osprey` throughout — the name change matches ProteoWizard/pwiz#4335 ("Renamed OspreySharp to Osprey"), so there is no longer an OspreySharp binary to look for. `grep -ri ospreysharp` over the tree returns nothing.
